### DISCUSSION TITLE
Throat diameters are much larger than pore diameters from snow extraction

### DISCUSSION THREAD

**Discussion by JoHommel** [PMEAL/porespy#548] [Q&A] [answered]
I am I am Johannes Hommel, a PostDoc at the University of Stuttgart, and more or less recently started using PoreSpy to extract pore networks from CT images of sintered glass bead columns in which calcite was precipitated and, more recently, 3D reconstructions of "quasi-2D" mircofluidic experiments, which were imaged by simple microscopy. Due to the precipitation over time, the initially identical images in the z-direction will become different as more and more calcite accumulates. This is also why I would like to have the resolution of the z-direction, reconstructed from the 2D microscopy images as fine as possible. However, concerning this email, everything I write relates to the "empty/clean" cells, so the resulting pore network should indeed be 2-dimensional and the height of all pore throats and pore bodies be simply the resolution of the reconstruction in z-direction.

In the "quasi-2D" setups, it became suddenly quite obvious, that the throat diameters as extracted by PoreSpy (I have the 1.3.1 version installed) seem off. Especially odd is that the throat inscribed diameter, which in a "quasi-2D" setup should be <= the height of the microfluidic cell, seems to vary greatly and often is 2 to 5 times the microfluidic cell's height. The throat equivalent diameter is also at times much larger than the microfluidic cell's height, but this is to be expected as some of the throats in the cell are often much wider than the microfluidic cell's height. Really strange is that, in general, the throat inscribed diameters are larger than the throat equivalent diameters, although I did not compare the values for all the few thousand  throats. 

I initially suspected that for some reasons, the created pore network is not completely flat (2D), but all the pore bodies have the same coordiantes in z direction, so I guess having throats with some angle to the x,y plane cannot be the reason for the large throat diameters. 

Interestingly, the the pore (body) diameters seem correct.

Additionally, I noted that when I increase the resolution of the reconstruction of the 2D microscopy in the 3rd direction, and thus also in the other two dimensions, the pore network gets oversegmented resulting in more than one "pore" in the network where I visually would assign only a single pore. I could somewhat, but not completely, compensate by trying to adjust the "r" and "sigma" parameters of the SNOW Algorithm.

I would greatly appreciate any hints at what to look at in detail to figure out if/when/where somethings goes wrong or any other advice. I can also provide some examples (geometries, csv tables with the extracted pore networks and pore/throat parameters, ...) as needed, but wanted to wait for if and what you exactly deem necessary.

**Comment by jgostick** [comment]
Interesting.  The inscribed diameter should certainly not exceed the model height, so your concerns are valid there.  I will have to investigate and get back to you.  Would you be able to send me a 2D image of our micromodel?  I can then pad it into 3D.

Regarding the over-segmentation, this is just a common problem with the watershed.  We have tried to reduce this as much as possible with our algorithm, but it's never going to be perfect for all cases.

**Reply by ma-sadeghi** [reply]
@JoHommel Also, try using the `dev` version. `v1.3.1` is outdated and we're in the process of releasing v2 soon. I don't think it solves your problem though, but it is worth a try. See [here](https://porespy.org/user_guide/installation.html#installing-the-dev-version) for instructions on how to install the `dev` version.

**Reply by JoHommel** [reply]
Thank you for the replies. 

I attached an image in the lowest resolution I have it here. For the empty microfluidic cell without precipitation at this resolution, I just used this image four times in z-direction as the x,y pixel size is 8µm and the total height is 30-35µm.
![2020_09_04_17-29-06-782_1](https://user-images.githubusercontent.com/87419764/125635971-899f1f4c-d9c2-4a2b-a91b-25fc05ea9806.png)

I'll check out dev, but v1.3.1 was the most recent version available for "easy" installation.

**Reply by jgostick** [reply]
For instance, you should do this:

``` python

import porespy as ps
import imageio
import numpy as np
import matplotlib.pyplot as plt

im = imageio.imread('micromodel.png')

im = np.tile(np.atleast_3d(im), reps=[1, 1, 10])
im[..., 0] = 0
im[..., -1] = 0

```

**Reply by JoHommel** [reply]
Thanks! I'll try that the next days. It sound like it is a simple solution and of course also adds the realistic information on the top and bottom of the cell. I never really thought into this direction. I'll get back, but probably not before Friday.

**Reply by kweishaupt** [reply]
I wanted to ask a similar question, but Johannes was faster :-)

I also tried to extract a network from a 2D png image where I noticed a lot of very small pore regions within the throats.
I could improve that a bit by setting `sigma` to really large values (4.0) but I still noticed some strange areas like the blue one here:
![image](https://user-images.githubusercontent.com/74908731/125649307-38d30a25-652e-4295-b917-e5a71459e861.png)

Is this a general issue with 2D images?

**Reply by jgostick** [reply]
I think that your image is basically TOO highly resolved.  The sigma and r_max arguments are basically telling the algorithm how far to look for peaks, yet your pore spaces are enormous, in terms of voxels.  

We have been playing with a new way to find peaks, that is not dependent on size...perhaps we should follow through with this project sooner than later?

**Reply by kweishaupt** [reply]
> your image is basically TOO highly resolved

That's interesting. The situation actually does become better if I decrease the resolution, thanks!
Is there maybe some rule of thump regarding the resolution? Maybe something like `pixel per pore diameter` or so?
We generate the images artificially.

**Reply by ma-sadeghi** [reply]
@JoHommel Fair enough, the instructions in our docs for installing the `dev` version are not "easy" (although they're the right way if you want to contribute to `porespy` or easily get new updates as they roll in without having to uninstall `porespy` every time). There's a simplified way, although not encouraged. Just run:
```bash
pip install git+https://github.com/PMEAL/porespy.git@dev
```
I also updated our [docs](https://porespy.org/user_guide/installation.html#installing-the-dev-version).

**Reply by jgostick** [reply]
@kweishaupt, I have been playing with my "alternative" method, and have found that by just putting sigma = 5.0, instead of 0.5, I get a pretty decent segementation.  In other words, when the pores are large and well resolved, the use of much larger sigma is required to smooth out the extraneous peaks.  If you put sigma=5.0, the smoothing will extend much further.  I was not surprized that this worked well in larger pores, but I was surprized that it didn't seem to hurt the segmentation of smaller pores even tough the blur kernal clearly extends well beyond the small pores.  So when I create a multiscale image and use sigma=5.0, all pores seem pretty well segmented.  Hope that helps.

**Reply by kweishaupt** [reply]
@jgostick That does help, thanks a lot. I was asking myself if it could be useful to expose the `sigma` and `r` params to the interface of `snow2(...)` such that it can be changed more easily.

**Reply by kweishaupt** [reply]
@jgostick I get pretty good results now on a 2D image with reduced resolution. If I apply the tiling as described above, however, boundary pores within the interior of the domain appear. Any hint where this could come from (does not happen in 2D)?

I used this file with sigma = 3, resolution = 1e-6 m.

![test](https://user-images.githubusercontent.com/74908731/125802202-390c5a75-ef0e-40d3-846a-8ef5832fe2a8.png)


Do I understand correctly that I could just take the network extracted from the 2D image and multiply volumes and areas by my given height in order to get the correct values? Of course I would also have to clip the diameters using min(diameter, height).

**Reply by jgostick** [reply]
I do not know why boundary pores would appear in the middle of the domain, and it seems like I don't quite understand what your describing.  My first suggestions would be 

(a) put more than 1 layer of solid voxels on the top and bottom, maybe do 3 layers of voxels, so:

``` python
im = np.tile(np.atleast_3d(im), reps=[1, 1, 15])  # Pad extra thick
im[..., 0:3] = 0  # Set first 3 layers to solid
im[..., -3:] = 0  # Set last 3 layers to solid

```

and (b) make sure that you're not adding boundary pores to the top and bottom faces (though if you have a thick padding of 0's on the top and bottom then the boundary pore process won't find any pores there).

**Reply by jgostick** [reply]
To answer your other question, no, do not do a 2D extraction and scale it.  2D extraction is not really physical and is only good for demonstrations.  The determination of pore properties from 2D images is tricky.  Of course everything is possible. If you get very good extractions in 2D, then you could manually compute the "3D" sizes knowing the diameters of inscribed spheres and such.

**Reply by kweishaupt** [reply]
Thanks! This is the extracted network using the extra thick padding:
![image](https://user-images.githubusercontent.com/74908731/125804818-569b6a30-751a-4510-afdb-6a6c9ba81e45.png)

You still get two boundary pores within the domain.

**Reply by kweishaupt** [reply]
If I set `boundary_width=[3,3,0]`, the internal boundary pores disappear. I guess this is what you meant?

**Reply by jgostick** [reply]
Yes, I meant exactly that: (3, 3, 0).  I'm glad that fixed your problem, but I don't know why it would find 2 boundary pores when set to (3,3,3)!  The solid walls should result in more solid when the domain is tiled.  Arg! The edge cases seem to never end.  :-(

**Reply by kweishaupt** [reply]
@jgostick Sorry for digging up these cases :-(

Shall I open an issue regarding the parameters sigma and r_max in the interface of `snow2(...)`?

**Reply by ma-sadeghi** [reply]
@kweishaupt That's already merged into `dev` in #549, pull the latest changes from `dev` and you can pass `sigma` and `r_max` to `snow2`.

**Reply by kweishaupt** [reply]
Perfect, thanks a lot!

**Reply by JoHommel** [reply]
@ma-sadeghi I managed to change from 1.31 to dev "The hard (but correct) way".
Although I thought I updated all the requirements/required packages, there were still some I had to upgrade manually and it took me some time being a bit unfamiliar with python.  However, now everything seems to work.

@jgostick The inscribed diameters are now as expected: <= cell height. Everything looks good so far. 

Thanks for the help!

**Reply by ma-sadeghi** [reply]
@JoHommel Do you remember which packages were causing the issue? Was it `docrep` by any chance?

**Reply by JoHommel** [reply]
The last was: tqdm , were I previously had version 4.42.1, which was not sufficient and now, with tqdm-4.61.2 porespy works. 

For the other packages, I fear I lost that information as I cannot scroll back far enough in my terminal anymore as I did that yesterday while attending some online meetings in parallel. They weren't that many, maybe 2 or 3?

**Comment by jgostick** [comment, marked as answer]
When you tile this image to make it 'thick', you need to put 0's on the top and bottom to represent the plates or glass covers or whatever you have there. Otherwise the distance transform will treat the z-direction as infinite, so the nearest solid will be the pillar, not the plates. This may also help with the watershed partitioning problem since the contours of the distance transform will be better defined.  The following code snippet explains how to do this:

``` python
import porespy as ps
import imageio
import numpy as np
import matplotlib.pyplot as plt

im = imageio.imread('micromodel.png')

im = np.tile(np.atleast_3d(im), reps=[1, 1, 10])
im[..., 0] = 0
im[..., -1] = 0
```
